### ISSUE TITLE
Getting KeyError after using "bins" parameter in Choropleth

### ISSUE THREAD

**Issue by emontarsolo** [python-visualization/folium#1254]
I tried to do a Choropleth map using Folium. The result was ok but I wanted the colormap to be more "shaded", since by default there is only 6 bins.
However when I set the parameter "bins=10", I get the error "KeyError: 'YlGn' "
I noticed I have the same error, when I specify the "bins" parameters in the "[Choropleth maps](https://python-visualization.github.io/folium/quickstart.html#Choropleth-maps)" example of the QuickStart. So I will use this example to illustrate my issue.

The following code:
```
import folium
import pandas as pd


url = 'https://raw.githubusercontent.com/python-visualization/folium/master/examples/data'
state_geo = f'{url}/us-states.json'
state_unemployment = f'{url}/US_Unemployment_Oct2012.csv'
state_data = pd.read_csv(state_unemployment)

m = folium.Map(location=[48, -102], zoom_start=3)

folium.Choropleth(
    geo_data=state_geo,
    name='choropleth',
    data=state_data,
    columns=['State', 'Unemployment'],
    key_on='feature.id',
    fill_color='YlGn',
    fill_opacity=0.7,
    line_opacity=0.2,
    legend_name='Unemployment Rate (%)',
    bins=10
).add_to(m)

folium.LayerControl().add_to(m)

m
```

Leads to this error :
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-6-d2bc56d78020> in <module>
     20     line_opacity=0.2,
     21     legend_name='Unemployment Rate (%)',
---> 22     bins=10
     23 ).add_to(m)
     24 

/usr/local/lib/python3.6/site-packages/folium/features.py in __init__(self, geo_data, data, columns, key_on, bins, fill_color, nan_fill_color, fill_opacity, nan_fill_opacity, line_color, line_weight, line_opacity, name, legend_name, overlay, control, show, topojson, smooth_factor, highlight, **kwargs)
   1088             # We add the colorscale
   1089             nb_bins = len(bin_edges) - 1
-> 1090             color_range = color_brewer(fill_color, n=nb_bins)
   1091             self.color_scale = StepColormap(
   1092                 color_range,

/usr/local/lib/python3.6/site-packages/branca/utilities.py in color_brewer(color_code, n)
    158     if not explicit_scheme:
    159         # Check to make sure that it is not a qualitative scheme.
--> 160         if scheme_info[base_code] == 'Qualitative':
    161             matching_quals = []
    162             for key in schemes:

KeyError: 'YlGn'
```

The same code without the " bins=10" works perfectly.

What Am I doing wrong here ? Is it because the [color_brewer schemes](https://github.com/python-visualization/folium/blob/v0.2.0/folium/utilities.py#L104) only take 6 values ? What Am I supposed to do then ?

Python version : 3.6.8
Folium Version : '0.10.1'

**Comment by Conengmo** [comment]
Weird. `bins` accepts values higher than 6, that shouldn't be a problem. The error is weird, `'YlGn'` isn't a key in `scheme_info`, though if you [look in the code](https://github.com/python-visualization/folium/blob/v0.2.0/folium/utilities.py#L112) that key is there. The part that does this lookup only is triggered when there are more than 6 bins, so that explains why you didn't get the error with 6 bins.

**Comment by emontarsolo** [comment]
@Conengmo Thank you for your answer.

Sorry I forgot to give an update ! I found a solution to get around this issue: I use the GeoJson feature instead of the Choropeth feature. Since GeoJson allows you to use custom colormap, and even a continuous colormap, you don't have to set a number of bins. This [example testing different colormaps](https://nbviewer.jupyter.org/github/python-visualization/folium/blob/v0.2.0/examples/Colormaps.ipynb) and also this [example using colormap + tooltip](https://nbviewer.jupyter.org/gist/jtbaker/57a37a14b90feeab7c67a687c398142c?flush_cache=true) help me a lot.

I'm not sure I should close the issue since the bug is still here ?

**Comment by Conengmo** [comment]
Hi @emontarsolo, thanks for reporting back, happy to hear you got it working using GeoJson. Let's leave the issue open, me or someone else should check if this bug indeed exists and in that case what should be fixed.

**Comment by ja-ox** [comment]
> Weird. `bins` accepts values higher than 6, that shouldn't be a problem. The error is weird, `'YlGn'` isn't a key in `scheme_info`, though if you [look in the code](https://github.com/python-visualization/folium/blob/v0.2.0/folium/utilities.py#L112) that key is there. The part that does this lookup only is triggered when there are more than 6 bins, so that explains why you didn't get the error with 6 bins.

The problem is in the `branca` `utilities.py`, not the `folium` one:

    /usr/local/lib/python3.6/site-packages/branca/utilities.py in color_brewer(color_code, n)

**Comment by Conengmo** [comment]
Hi @ja-ox, thanks for looking into this! You're right, this is currently broken in branca. 

Here we use the palette name and the number of bins to create a new name, which may not exist: https://github.com/python-visualization/branca/blob/master/branca/utilities.py#L128 We should make sure the integer we attach to the name results in an existing name.

A second problem is on this line: https://github.com/python-visualization/branca/blob/master/branca/utilities.py#L152 We should make this lookup robust, i.e. it doesn't matter if it's not in that dictionary, because then it's not `Qualitative`.

Is this something you want to help with? Or were you just pointing it out?

**Comment by ishant001** [comment]
> I tried to do a Choropleth map using Folium. The result was ok but I wanted the colormap to be more "shaded", since by default there is only 6 bins.
> However when I set the parameter "bins=10", I get the error "KeyError: 'YlGn' "
> I noticed I have the same error, when I specify the "bins" parameters in the "[Choropleth maps](https://python-visualization.github.io/folium/quickstart.html#Choropleth-maps)" example of the QuickStart. So I will use this example to illustrate my issue.
> 
> The following code:
> 
> ```
> import folium
> import pandas as pd
> 
> 
> url = 'https://raw.githubusercontent.com/python-visualization/folium/master/examples/data'
> state_geo = f'{url}/us-states.json'
> state_unemployment = f'{url}/US_Unemployment_Oct2012.csv'
> state_data = pd.read_csv(state_unemployment)
> 
> m = folium.Map(location=[48, -102], zoom_start=3)
> 
> folium.Choropleth(
>     geo_data=state_geo,
>     name='choropleth',
>     data=state_data,
>     columns=['State', 'Unemployment'],
>     key_on='feature.id',
>     fill_color='YlGn',
>     fill_opacity=0.7,
>     line_opacity=0.2,
>     legend_name='Unemployment Rate (%)',
>     bins=10
> ).add_to(m)
> 
> folium.LayerControl().add_to(m)
> 
> m
> ```
> 
> Leads to this error :
> 
> ```
> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> <ipython-input-6-d2bc56d78020> in <module>
>      20     line_opacity=0.2,
>      21     legend_name='Unemployment Rate (%)',
> ---> 22     bins=10
>      23 ).add_to(m)
>      24 
> 
> /usr/local/lib/python3.6/site-packages/folium/features.py in __init__(self, geo_data, data, columns, key_on, bins, fill_color, nan_fill_color, fill_opacity, nan_fill_opacity, line_color, line_weight, line_opacity, name, legend_name, overlay, control, show, topojson, smooth_factor, highlight, **kwargs)
>    1088             # We add the colorscale
>    1089             nb_bins = len(bin_edges) - 1
> -> 1090             color_range = color_brewer(fill_color, n=nb_bins)
>    1091             self.color_scale = StepColormap(
>    1092                 color_range,
> 
> /usr/local/lib/python3.6/site-packages/branca/utilities.py in color_brewer(color_code, n)
>     158     if not explicit_scheme:
>     159         # Check to make sure that it is not a qualitative scheme.
> --> 160         if scheme_info[base_code] == 'Qualitative':
>     161             matching_quals = []
>     162             for key in schemes:
> 
> KeyError: 'YlGn'
> ```
> 
> The same code without the " bins=10" works perfectly.
> 
> What Am I doing wrong here ? Is it because the [color_brewer schemes](https://github.com/python-visualization/folium/blob/v0.2.0/folium/utilities.py#L104) only take 6 values ? What Am I supposed to do then ?
> 
> Python version : 3.6.8
> Folium Version : '0.10.1'

make bins=9
color YlGn has 9 shades

**Comment by lorenzoromani1983** [comment]
Hi. Is this fixed? I can only plot maps up to 9 bins. More than 9 bins still gives keyerror.

**Comment by Conengmo** [comment]
AFAIK this hasn't been fixed yet. We need a volunteer to work on this.

**Comment by Conengmo** [comment]
I believe this was fixed with https://github.com/python-visualization/branca/pull/115, which was included in branca 0.6.0.
